containerd: 1.7.23 -> 2.0.0

getchoo opened the pull request:
Closes #356474

Diff: https://github.com/containerd/containerd/compare/v1.7.23...v2.0.0

Changelog: https://github.com/containerd/containerd/releases/tag/v2.0.0

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

getchoo committed:
containerd: format with nixfmt

getchoo committed:
containerd: 1.7.23 -> 2.0.0

This release contains breaking changes. See the full [containerd
2.0](https://github.com/containerd/containerd/blob/main/docs/containerd-2.0.md)
documentation for more

Diff: https://github.com/containerd/containerd/compare/v1.7.23...v2.0.0

Changelog: https://github.com/containerd/containerd/releases/tag/v2.0.0

getchoo committed:
containerd: use standard attributes for make

Instead of inlining flags and targets, we can use standard attributes
and emulate `mkDerivations`'s default build and install phases to make
maintaining and overriding the expression much easier

getchoo committed:
containerd: use best practices

getchoo committed:
containerd: add updateScript

getchoo committed:
containerd: add override for btrfs support

Having an explicit attribute to override for btrfs support is much more
user friendly

getchoo committed:
containerd: split outputs

getchoo committed:
containerd: add meta.mainProgram

getchoo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356618`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package marked as broken and skipped:</summary>
  <ul>
    <li>docker_24</li>
  </ul>
</details>
<details>
  <summary>:x: 16 packages failed to build:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>devcontainer</li>
    <li>docker (docker_27)</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>fn-cli</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
    <li>tests.devShellTools.nixos</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>containerd</li>
    <li>containerd.doc</li>
    <li>containerd.man</li>
    <li>flintlock</li>
    <li>nomad-driver-containerd</li>
  </ul>
</details>

Looks like `docker` [reuses our expression](https://github.com/NixOS/nixpkgs/blob/dae6c9d29ee93dc605a45d073120cf0c52543723/pkgs/applications/virtualization/docker/default.nix#L52-L65) for v1, so we'll need to keep that around still

getchoo committed:
docker: only install containerd binaries

Previously this override was broken as it was trying to use `make
install-doc` in a v1 source, while that target was introduced in v2.
This is fixed by only running `make install` for the binaries -- as
documentation is pointless in this build anyways

getchoo committed:
containerd: add getchoo to maintainers

getchoo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356618`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>containerd</li>
    <li>containerd.doc</li>
    <li>containerd.man</li>
    <li>devcontainer</li>
    <li>docker</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>flintlock</li>
    <li>fn-cli</li>
    <li>nomad-driver-containerd</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>

Kamillaova commented on pkgs/by-name/co/containerd/package.nix:
```diff
@@ -1,56 +1,82 @@
+
+  postConfigure = ''
+    export GOCACHE=$TMPDIR/go-cache
+    export GOPATH="$TMPDIR/go"
```
I'm not 100% sure that `GOPATH` is needed here, but

I think this might be useful:
```nix
disallowedReferences = [ go ];
inherit (go) GOOS GOARCH;
```

Kamillaova commented on pkgs/by-name/co/containerd/package.nix:
```diff
@@ -1,56 +1,85 @@
+    util-linux
+  ];
 
   buildInputs = [ btrfs-progs ];
```
```suggestion
  buildInputs = lib.optional (!btrfsSupport) btrfs-progs;
```

This will exclude `btrfs-progs` from the build closure when user explicitly sets `btrfsSupport = false`

Kamillaova replied:
Or maybe just like that, if you're supposed to disable btrfs using `override { btrfs-progs = null; }`

```patch
-  tags = lib.optional (!btrfsSupport) "no_btrfs";
+  tags = lib.optional (btrfs-progs != null) "no_btrfs";
```

Kamillaova reviewed (changes requested):
Are you sure that `btrfs-progs` in `buildInputs` is needed here?

Also it seems that (upcoming) runc 1.2 now can communicate to systemd via dbus, and because of that upstream made this change: https://github.com/containerd/containerd/commit/18e4ea9a6c5378a175e5090ca4052d0b2ff742df (landed in 2.0-rc1)
I think we need to do this too?

and... cross-compilation seems to be broken (and maybe native riscv64-linux too)

`aarch64-linux` -> `riscv64-linux` -> `riscv64-linux`
```
> Running phase: buildPhase
> build flags: SHELL=/nix/store/4vvkmiq95x6zzz82lcsbq8607x5v7aws-bash-5.2p37/bin/bash PREFIX=/nix/store/s6rwnryf14wzwc2yp8n9z4yn5akm223r-containerd-riscv64…
> + bin/ctr
> go build  -gcflags=-trimpath=/build/go/src -buildmode=pie  -o bin/ctr -ldflags '-X github.com/containerd/containerd/v2/version.Version=v2.0.0 -X github.c…
> -buildmode=pie requires external (cgo) linking, but cgo is not enabled
> make: *** [Makefile:264: bin/ctr] Error 1
For full logs, run 'nix log /nix/store/92fahvxcw4fqc6qgr3f1w9q8jqdri9j4-containerd-riscv64-unknown-linux-gnu-2.0.0.drv'.
```

Adding `CGO_ENABLED = 1;` will fix this error (but not manpages generation)

$out/bin/containerd-shim-runc-v2 remains nocgo, because crd's makefiles explicitly sets `CGO_ENABLED=0` for some cmds

Also that is strange why `CGO_ENABLED` is `0` on riscv64, but it is `1` on `x86_64-linux` — guesses, maybe not so

But
`aarch64-linux` -> `x86_64-linux` -> `x86_64-linux`
build fails with....

```
> Running phase: installPhase
> install flags: SHELL=/nix/store/4vvkmiq95x6zzz82lcsbq8607x5v7aws-bash-5.2p37/bin/bash PREFIX=/nix/store/xcjz29929cisqgy1vg1wwl56nspywyzc-containerd-x86_6…
> + install bin/ctr bin/containerd bin/containerd-stress bin/containerd-shim-runc-v2
> + install-doc
> + bin/gen-manpages
> go build  -gcflags=-trimpath=/build/go/src -buildmode=pie  -o bin/gen-manpages -ldflags '-X github.com/containerd/containerd/v2/version.Version=v2.0.0 -X…
> + man/ctr.8
> bin/gen-manpages ctr.8 man
> [qemu]: Could not open '/lib64/ld-linux-x86-64.so.2': No such file or directory
> make: *** [Makefile:293: man/ctr.8] Error 255
For full logs, run 'nix log /nix/store/gvysvl38jvway8h80gv39hrhzqj4jwvh-containerd-x86_64-unknown-linux-gnu-2.0.0.drv'.
```

BUT it builded the `ctr` cmd without `CGO_ENABLED=1` - what?

and yes, it tries to run a native binary when cross compiling (I'm using orbstack, qemu comes from it), and also it successfully builds after adding `CGO_ENABLED=1` - what?

so this is 100% not going to be built without binfmt_misc
I think this can be fixed in three ways:
- use emulatorFor if buildPlatform cannot execute hostPlatform
- compile `gen-manpages` with `GOOS`, `GOARCH` from `pkgsBuildBuild.go`
- disable manpages when cross compiling for a "local unexecutable" platform

It is even worse because orbstack also adds binfmt_misc for riscv64 and I can't disable it lol, so I can't fully test cross compilation right now

Kamillaova commented:
Native `aarch64-linux` build now also fails, but due to `disallowedReferences`

adding `GOFLAGS = [ "-trimpath" ];` will fix it, but why did this work without it when cross-compiling?

and also - what is this lol?
https://github.com/containerd/containerd/blob/a4579956c0c951061dc86eef98029a9e29a65828/Makefile#L139

Kamillaova commented:
Okay I disabled all emulators (https://github.com/orbstack/orbstack/issues/1592), and as expected it fails with this: `bin/gen-manpages: cannot execute binary file: Exec format error`

Kamillaova commented:
wait, `rebuild-darwin: 1`, what?

getchoo committed:
containerd: conditionally build manpages

An in-tree binary is used to build some manpages, which won't work when
cross compiling unless the buildPlatform can execute code for the
hostPlatform

getchoo committed:
containerd: add cross compilation tests

getchoo committed:
nixos/containerd: load after `local-fs.target` & `dbus.service`

https://github.com/containerd/containerd/commit/18e4ea9a6c5378a175e5090ca4052d0b2ff742df

getchoo commented:
Cross compilation issues have been fixed by moving back to buildGoModule -- they're also tested on both Darwin and Linux now!

> Are you sure that btrfs-progs in buildInputs is needed here?

[Yes](https://github.com/containerd/containerd/blob/6b6f53cb1e0bac29d9fc8792266ed56a684a012e/BUILDING.md#build-requirements)

> wait, rebuild-darwin: 1, what?

[According to ofborg](https://gist.github.com/GrahamcOfBorg/49100e2a7e306c95602ef2afca338217), this changes `tests.devShellTools.nixos` on aarch64/x86_64-darwin 🤷🏻

getchoo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356618`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>containerd</li>
    <li>containerd.doc</li>
    <li>containerd.man</li>
    <li>devcontainer</li>
    <li>docker</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>flintlock</li>
    <li>fn-cli</li>
    <li>nomad-driver-containerd</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>

Kamillaova reviewed (commented):
> Cross compilation issues have been fixed by moving back to buildGoModule

Wait, what? Or do you mean there is no need to manually set golang-specific attributes anymore? Like `-trimpath`

> they're also tested on both Darwin and Linux now!

I really don't think anyone will try to build containerd with darwin->linux cross, and currently there is no triggers for building darwin->linux cross stdenv and packages in hydra

Am I right in understanding that `containerd.tests.cross-amd64.{x86_64,aarch64}-darwin` attributes will be built in hydra? If so, that's not sounds cool, because it will just waste cpu time of darwin builders, storage of the cache.nixos.org... for packages that no one will use

About aarch64-linux on x86_64-linux... okay?


And I think (not 100% sure) you can guard the `go-md2man` package in `buildInputs` with `lib.optional withMan`

---
UPD: Ah I understand, `<drv>.tests` doesn't run in hydra, but I think darwin->linux cross test are not needed.

Also, maybe add cross-tests like this? https://github.com/NixOS/nixpkgs/blob/nixos-24.05/pkgs/os-specific/linux/systemd/default.nix#L884

Kamillaova commented on pkgs/by-name/co/containerd/package.nix:
```diff
@@ -1,56 +1,103 @@
-  nativeBuildInputs = [ go-md2man installShellFiles util-linux ];
+  strictDeps = true;
+
+  nativeBuildInputs = [
```
`nativeBuildInputs = [ util-linux ] ++ lib.optional withMan go-md2man;`

getchoo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356618`

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>airlift</li>
    <li>airlift.dist</li>
    <li>charliecloud</li>
    <li>containerd</li>
    <li>containerd.doc</li>
    <li>containerd.man</li>
    <li>devcontainer</li>
    <li>docker</li>
    <li>docker-gc</li>
    <li>docker-sbom</li>
    <li>docker_25</li>
    <li>docker_26</li>
    <li>flintlock</li>
    <li>fn-cli</li>
    <li>nomad-driver-containerd</li>
    <li>pipework</li>
    <li>python311Packages.jupyter-repo2docker</li>
    <li>python311Packages.jupyter-repo2docker.dist</li>
    <li>python312Packages.jupyter-repo2docker</li>
    <li>python312Packages.jupyter-repo2docker.dist</li>
  </ul>
</details>

Kamillaova commented:
@getchoo `CONFLICT (content): Merge conflict in nixos/doc/manual/release-notes/rl-2505.section.md`
